File: android/app/src/main/kotlin/com/llloooggg/letsflutssh/SecureTextPlugin.kt
package com.llloooggg.letsflutssh

import android.content.Context
import android.graphics.Color
import android.text.InputType
import android.util.TypedValue
import android.view.View
import android.view.inputmethod.EditorInfo
import android.widget.EditText
import android.widget.FrameLayout
import androidx.core.widget.addTextChangedListener
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.StandardMessageCodec
import io.flutter.plugin.platform.PlatformView
import io.flutter.plugin.platform.PlatformViewFactory
import java.nio.charset.StandardCharsets

/// Native-memory-backed secure text input for password / passphrase
/// fields. Bytes live in an Android [EditText]'s [android.text.Editable]
/// (Java heap, char[] — mutable) instead of a Dart [String] (Dart
/// heap, immutable, GC-relocatable). On submit the bytes are encoded
/// to UTF-8, sent across the method channel as a [Uint8List], and
/// the EditText content is wiped in place before the call returns.
/// The Dart side copies the bytes into a page-locked [SecretBuffer]
/// and zeroes the Uint8List in the same frame — residency on the
/// Dart heap collapses to a single copy, one frame long.
///
/// Caveat: the bytes still sit in the Java heap for the duration of
/// typing. Java heap is also GC'd, but Android keyboards (Samsung,
/// Gboard, SwiftKey, Chinese IMEs) already handle passwords through
/// this same EditText path — the threat model is narrower than
/// "every process can read it". The win over Flutter TextField is
/// eliminating the interim Dart [String] allocations that land per
/// keystroke in the TextEditingValue history.
class SecureTextFactory(private val messenger: BinaryMessenger) :
    PlatformViewFactory(StandardMessageCodec.INSTANCE) {
    override fun create(context: Context, viewId: Int, args: Any?): PlatformView {
        return SecureTextView(context, messenger, viewId)
    }
}

class SecureTextView(
    context: Context,
    messenger: BinaryMessenger,
    viewId: Int,
) : PlatformView {
    companion object {
        const val VIEW_TYPE = "com.letsflutssh/secure_text"
    }

    private val channel = MethodChannel(
        messenger,
        "com.letsflutssh/secure_text_$viewId",
    )

    private val editText = EditText(context).apply {
        inputType = InputType.TYPE_CLASS_TEXT or
            InputType.TYPE_TEXT_VARIATION_PASSWORD
        imeOptions = EditorInfo.IME_ACTION_DONE or
            EditorInfo.IME_FLAG_NO_PERSONALIZED_LEARNING or
            EditorInfo.IME_FLAG_NO_EXTRACT_UI
        isSingleLine = true
        isLongClickable = false
        setTextIsSelectable(false)
        setTextColor(Color.BLACK)
        setHintTextColor(Color.GRAY)
        setTextSize(TypedValue.COMPLEX_UNIT_SP, 14f)
    }

    private val container = FrameLayout(context).apply {
        addView(
            editText,
            FrameLayout.LayoutParams(
                FrameLayout.LayoutParams.MATCH_PARENT,
                FrameLayout.LayoutParams.WRAP_CONTENT,
            ),
        )
    }

    init {
        editText.addTextChangedListener(
            afterTextChanged = { _ ->
                // Notify Dart when there's any text vs. empty — the
                // Dart side uses this to light up the Apply button.
                channel.invokeMethod(
                    "onChanged",
                    mapOf("hasText" to (editText.text?.isNotEmpty() == true)),
                )
            },
        )
        editText.setOnEditorActionListener { _, actionId, _ ->
            if (actionId == EditorInfo.IME_ACTION_DONE) {
                emitAndWipe()
                true
            } else {
                false
            }
        }
        channel.setMethodCallHandler { call, result ->
            when (call.method) {
                "submit" -> {
                    val bytes = emitAndWipe()
                    result.success(bytes)
                }
                "focus" -> {
                    editText.requestFocus()
                    result.success(true)
                }
                "clear" -> {
                    wipeEditable()
                    result.success(true)
                }
                else -> result.notImplemented()
            }
        }
    }

    /// Encode current content to UTF-8 and wipe the Editable in
    /// place. Returns the bytes by value so the native→Dart channel
    /// can deliver them even after the view is gone.
    private fun emitAndWipe(): ByteArray {
        val editable = editText.text ?: return ByteArray(0)
        val bytes = editable.toString().toByteArray(StandardCharsets.UTF_8)
        // Also push the bytes through the one-shot onSubmit stream so
        // a long-press context-menu "Done" path surfaces the same
        // event shape as the IME Done key.
        channel.invokeMethod("onSubmit", bytes)
        wipeEditable()
        return bytes
    }

    /// Overwrite the backing char[] with NUL before clearing. The
    /// default `setText("")` replaces the underlying array without
    /// zeroing the old one — ART may hold onto the deallocated
    /// char[] in the young generation until the next GC cycle.
    /// Overwriting forces the same bytes to be NUL regardless of
    /// GC timing.
    private fun wipeEditable() {
        val editable = editText.text ?: return
        val len = editable.length
        if (len > 0) {
            val nulls = CharArray(len)
            editable.replace(0, len, String(nulls))
        }
        editable.clear()
    }

    override fun getView(): View = container

    override fun dispose() {
        wipeEditable()
        channel.setMethodCallHandler(null)
    }
}

/// Thin registrar — call from MainActivity.configureFlutterEngine.
class SecureTextPlugin {
    fun register(
        messenger: BinaryMessenger,
        registry: io.flutter.plugin.platform.PlatformViewRegistry,
    ) {
        registry.registerViewFactory(
            SecureTextView.VIEW_TYPE,
            SecureTextFactory(messenger),
        )
    }
}

File: ios/Runner/SecureTextPlugin.swift
import Flutter
import UIKit

/// Native-memory-backed secure text field for iOS. Sibling to the
/// Android `SecureTextPlugin` — same contract, same channel names,
/// same one-shot `submit` emission. See
/// `lib/widgets/secure_native_text_field.dart` for the rationale.
///
/// Typed bytes live in the `UITextField` `.text` (Swift `String` —
/// value type, still heap-allocated but mutable and explicitly
/// zero-able). On IME Return the bytes are UTF-8 encoded, delivered
/// to Dart as a mutable `FlutterStandardTypedData(.bytes)`, and the
/// text field content is wiped in place before the call returns.
///
/// The trade-off vs Flutter `TextField`: Dart-heap residency
/// collapses from many per-keystroke String allocations to a single
/// one-frame Uint8List. Swift-heap residency (the actual typing
/// buffer) is still there, but Swift `String` backing is mutable —
/// the widget overwrites it with NUL before clearing, the same
/// discipline Android's EditText wipe follows.
final class SecureTextPlugin {
    static let viewType = "com.letsflutssh/secure_text"

    private let messenger: FlutterBinaryMessenger
    private var factory: SecureTextFactory?

    init(messenger: FlutterBinaryMessenger) {
        self.messenger = messenger
    }

    func register(with registrar: FlutterPluginRegistrar) {
        let factory = SecureTextFactory(messenger: messenger)
        registrar.register(factory, withId: SecureTextPlugin.viewType)
        self.factory = factory
    }
}

final class SecureTextFactory: NSObject, FlutterPlatformViewFactory {
    private let messenger: FlutterBinaryMessenger

    init(messenger: FlutterBinaryMessenger) {
        self.messenger = messenger
        super.init()
    }

    func create(
        withFrame frame: CGRect,
        viewIdentifier viewId: Int64,
        arguments args: Any?
    ) -> FlutterPlatformView {
        return SecureTextView(
            frame: frame,
            viewId: viewId,
            messenger: messenger
        )
    }

    func createArgsCodec() -> FlutterMessageCodec & NSObjectProtocol {
        return FlutterStandardMessageCodec.sharedInstance()
    }
}

final class SecureTextView: NSObject, FlutterPlatformView, UITextFieldDelegate {
    private let container: UIView
    private let textField: UITextField
    private let channel: FlutterMethodChannel

    init(frame: CGRect, viewId: Int64, messenger: FlutterBinaryMessenger) {
        self.container = UIView(frame: frame)
        self.textField = UITextField(frame: container.bounds)
        self.channel = FlutterMethodChannel(
            name: "com.letsflutssh/secure_text_\(viewId)",
            binaryMessenger: messenger
        )
        super.init()

        textField.isSecureTextEntry = true
        textField.autocorrectionType = .no
        textField.autocapitalizationType = .none
        textField.spellCheckingType = .no
        textField.smartDashesType = .no
        textField.smartQuotesType = .no
        textField.smartInsertDeleteType = .no
        textField.textContentType = .password
        textField.returnKeyType = .done
        textField.delegate = self
        textField.addTarget(
            self,
            action: #selector(onEditingChanged),
            for: .editingChanged
        )
        textField.autoresizingMask = [.flexibleWidth, .flexibleHeight]
        textField.borderStyle = .none
        textField.backgroundColor = .clear
        textField.textColor = .black
        container.addSubview(textField)

        channel.setMethodCallHandler { [weak self] call, result in
            self?.handle(call: call, result: result)
        }
    }

    func view() -> UIView { container }

    // MARK: - Channel methods

    private func handle(call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "submit":
            let bytes = emitAndWipe()
            result(FlutterStandardTypedData(bytes: bytes))
        case "focus":
            textField.becomeFirstResponder()
            result(true)
        case "clear":
            wipeField()
            result(true)
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    // MARK: - UITextFieldDelegate

    func textFieldShouldReturn(_ textField: UITextField) -> Bool {
        let bytes = emitAndWipe()
        channel.invokeMethod(
            "onSubmit",
            arguments: FlutterStandardTypedData(bytes: bytes)
        )
        textField.resignFirstResponder()
        return true
    }

    @objc private func onEditingChanged(_ sender: UITextField) {
        let hasText = !(sender.text?.isEmpty ?? true)
        channel.invokeMethod("onChanged", arguments: ["hasText": hasText])
    }

    // MARK: - Wipe

    /// UTF-8 encode current content + overwrite the backing String
    /// with NUL before clearing. Swift `String` is a value type
    /// backed by a heap-allocated `_StringStorage`; overwriting
    /// produces the same NUL bytes in whichever arena the value
    /// currently resides in, so the GC-equivalent reclamation
    /// cannot resurrect the original characters.
    private func emitAndWipe() -> Data {
        let text = textField.text ?? ""
        let bytes = Data(text.utf8)
        wipeField()
        return bytes
    }

    private func wipeField() {
        if let t = textField.text, !t.isEmpty {
            let nulls = String(repeating: "\0", count: t.count)
            textField.text = nulls
        }
        textField.text = ""
    }
}
